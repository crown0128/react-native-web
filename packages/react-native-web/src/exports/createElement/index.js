/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import AccessibilityUtil from '../../modules/AccessibilityUtil';
import createDOMProps from '../../modules/createDOMProps';
import React from 'react';

const adjustProps = domProps => {
  const { onClick, role } = domProps;

  const isButtonLikeRole = AccessibilityUtil.buttonLikeRoles[role];
  const isDisabled = AccessibilityUtil.isDisabled(domProps);

  // Button-like roles should not trigger 'onClick' if they are disabled.
  if (isButtonLikeRole && isDisabled && domProps.onClick != null) {
    domProps.onClick = undefined;
  }

  // Button-like roles should trigger 'onClick' if SPACE or ENTER keys are pressed.
  if (isButtonLikeRole && !isDisabled) {
    domProps.onKeyPress = function(e) {
      if (!e.isDefaultPrevented() && (e.which === 13 || e.which === 32)) {
        e.preventDefault();
        if (onClick) {
          onClick(e);
        }
      }
    };
  }
};

const createElement = (component, props, ...children) => {
  // use equivalent platform elements where possible
  let accessibilityComponent;
  if (component && component.constructor === String) {
    accessibilityComponent = AccessibilityUtil.propsToAccessibilityComponent(props);
  }
  const Component = accessibilityComponent || component;
  const domProps = createDOMProps(Component, props);
  adjustProps(domProps);
  return React.createElement(Component, domProps, ...children);
};

export default createElement;
