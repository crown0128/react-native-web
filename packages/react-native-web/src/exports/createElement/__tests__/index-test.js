/* eslint-env jasmine, jest */

import createElement from '..';
import React from 'react';
import { render, shallow } from 'enzyme';

describe('modules/createElement', () => {
  test('renders different DOM elements', () => {
    let component = render(createElement('span'));
    expect(component).toMatchSnapshot();
    component = render(createElement('main'));
    expect(component).toMatchSnapshot();
  });

  describe('prop "accessibilityRole"', () => {
    test('and string component type', () => {
      const component = shallow(createElement('span', { accessibilityRole: 'link' }));
      expect(component.find('a').length).toBe(1);
    });

    test('and function component type', () => {
      const Custom = () => <div />;
      const component = shallow(createElement(Custom, { accessibilityRole: 'link' }));
      expect(component.find('div').length).toBe(1);
    });

    const testRole = ({ accessibilityRole, disabled }) => {
      [{ key: 'Enter' }, { key: ' ' }].forEach(({ key }) => {
        test(`"onClick" is ${disabled ? 'not ' : ''}called when "${key}" key is pressed`, () => {
          const onClick = jest.fn();
          const component = shallow(
            createElement('span', { accessibilityRole, disabled, onClick })
          );
          component.find('span').simulate('keyPress', {
            isDefaultPrevented() {},
            nativeEvent: {},
            preventDefault() {},
            key
          });
          expect(onClick).toHaveBeenCalledTimes(disabled ? 0 : 1);
        });
      });
    };

    describe('value is "button" and disabled is "true"', () => {
      testRole({ accessibilityRole: 'button', disabled: true });
    });

    describe('value is "button" and disabled is "false"', () => {
      testRole({ accessibilityRole: 'button', disabled: false });
    });

    describe('value is "menuitem" and disabled is "true"', () => {
      testRole({ accessibilityRole: 'menuitem', disabled: true });
    });

    describe('value is "menuitem" and disabled is "false"', () => {
      testRole({ accessibilityRole: 'menuitem', disabled: false });
    });
  });
});
