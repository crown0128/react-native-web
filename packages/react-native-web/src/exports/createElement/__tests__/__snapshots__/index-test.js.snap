// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/createElement renders different DOM elements 1`] = `<span />`;

exports[`modules/createElement renders different DOM elements 2`] = `<main />`;
